src/Storages/Streaming/StorageRandom.cpp

```diff
@@ -370,13 +372,21 @@ class GenerateRandomSource final : public ISource
         , context(context_)
         , events_per_second(events_per_second_)
         , header_chunk(Nested::flatten(block_full.cloneEmpty()).getColumns(), 0)
+        , generate_interval(interval_time_)
     {
         is_streaming = true;
-
-        boundary_time = MonotonicMilliseconds::now() + generate_interval;
         block_idx_in_window = 0;
         max_full_block_count = events_per_second_ / block_size;
         partial_size = events_per_second_ % block_size;
+        /// calculate the number of interval during 1 sec
+        interval_count = 1000 / generate_interval;
+        /// special interval
+        last_interval_time = generate_interval + 1000 % generate_interval;
+        boundary_time = MonotonicMilliseconds::now() + last_interval_time;
+        /// calculate the number of data generated per interval
+        normal_interval = events_per_second / interval_count;
+        /// calculate the number of data generated in the special interval
+        last_interval_count = normal_interval + events_per_second % interval_count;
```

Comment:
It's complex code here, we should have a clearer comments, and add a example to explain this logic.